### PR TITLE
Support commonjs npm packages in devmode

### PR DESCRIPTION
**Pull request by gregmagolan**
Devmode rules such as `ts_devserver` & `ts_web_test_suite` don't support commonjs npm packages as they require a named-UMD or named-AMD bundle.

The current work-around for this in angular-bazel-example for an npm package such as `date-fns` is to generate a named-UMD date-fns bundle using browserify in the postinstall step:

```
    "scripts": {
        ...
        "postinstall": "ngc -p angular-metadata.tsconfig.json && yarn browserify-date-fns",
        "// browserify-date-fns": "for ts_devserver & ts_web_test_suite, create named UMD bundle of date-fns npm package with AMD name 'date-fns' and global name browserifyDateFns",
        "browserify-date-fns": "browserify -p [ named-amd --name date-fns ] node_modules/date-fns/index.js -s browserifyDateFns -o node_modules/date-fns/date-fns.umd.js",
        ...
```

This PR attempts to improve this situation. Currently it adds a target that runs this same browserify step for each npm package. For date-fns this target is `@npm//node_modules/date-fns:date-fns__umd` which outputs `bazel-bin/external/npm/node_modules/date-fns/date-fns.umd.js`.

### CODE REVIEW

File: internal/npm_install/npm_amd_bundle.bzl [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+def _npm_amd_bundle(ctx):
+    if len(ctx.attr.entry_point.files) != 1:
+        fail("labels in entry_point must contain exactly one file")
+
+    output = ctx.actions.declare_file("%s.amd.js" % ctx.attr.package_name)
+
+    args = ctx.actions.args()
+
+    # package name
```
Comment (alexeagle) [inline internal/npm_install/npm_amd_bundle.bzl]: nit: these comments don't seem helpful

File: third_party/github.com/browserify/browserify/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Vendored browserify
+
+This directory contains the [browserify](https://github.com/browserify/browserify) npm package compiled into a single file by [@zeit/ncc](https://github.com/zeit/ncc) and minified by [terser](https://github.com/terser-js/terser.
```
Comment (alexeagle) [inline third_party/github.com/browserify/browserify/README.md]: missing closing paren

File: third_party/github.com/browserify/browserify/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,32 @@
+load("@build_bazel_rules_nodejs//:defs.bzl", "nodejs_binary")
+
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])
```
Comment (alexeagle) [inline third_party/github.com/browserify/browserify/BUILD.bazel]: include a comment saying where the files came from (including the tag/sha)

File: third_party/github.com/jhermsmeier/browserify-named-amd/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])
+
+exports_files(["LICENSE"])
```
Comment (alexeagle) [inline third_party/github.com/jhermsmeier/browserify-named-amd/BUILD.bazel]: comment saying where it came from
see https://github.com/angular/angular/blob/master/third_party/README.md

File: internal/npm_install/browserify-wrapped.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    // browserify will camelCase this name to browserifyFooBar
+    '-s', `browserify-${packageName}`,
+    // output
+    '-o', output
+  ];
+
+  if (DEBUG) console.error(`\nRunning: node ${args.join(' ')}\n`);
+
+  const isWindows = /^win/i.test(process.platform);
+  child_process.execFileSync(
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/npm_install/browserify-wrapped.js]: Browserify a node cli program which has `#!/usr/bin/env node` at the top and expects command line arguments. We use this pattern in `terser-wrapped.js` and `tsc-directory.js` as well. Browserify also implements its own node resolution so there it doesn't need the overloaded resolve. Not sure if there is a better way.

File: internal/npm_install/browserify-wrapped.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    // entry point
+    entryPoint,
+    // add named-amd plugin which is vendored in
+    '-p', '[',
+    require.resolve(
+        'build_bazel_rules_nodejs/third_party/github.com/jhermsmeier/browserify-named-amd/named-amd.js'),
+    '--name', packageName, ']',
+    // set global name to browserify-foo-bar so there are no conflicts with other globals;
+    // browserify will camelCase this name to browserifyFooBar
+    '-s', `browserify-${packageName}`,
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/npm_install/browserify-wrapped.js]: sure. doesn't hurt.

File: internal/npm_install/npm_amd_bundle.bzl (hunk 1 above)
Comment (gregmagolan) [reply internal/npm_install/npm_amd_bundle.bzl]: 🐼

File: third_party/github.com/browserify/browserify/README.md (hunk 2 above)
Comment (gregmagolan) [reply third_party/github.com/browserify/browserify/README.md]: 👍

File: internal/npm_install/generate_build_file.js [hunk 7]
```diff
@@ -789,6 +819,38 @@ node_module_library(
...
+  let mainEntryPoint = pkg.main ? cleanupEntryPointPath(pkg.main) : undefined;
+  if (mainEntryPoint) {
+    // check if main entry point exists
+    mainEntryPoint = findFile(pkg, mainEntryPoint) || findFile(pkg, `${mainEntryPoint}.js`);
+    if (!mainEntryPoint) {
+      // If "main" entry point listed could not be resolved to a file
+      // then don't create an npm_amd_bundle target. This can happen
+      // in some npm packages that list an incorrect main such as v8-coverage@1.0.8
+      // which lists `"main": "index.js"` but that file does not exist.
+      console.warn(`Could not find "main" entry point ${pkg.main} in npm package ${pkg._name}`);
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/npm_install/generate_build_file.js]: Not much point for the majority of users. I'll put the console.log behind a DEBUG flag.

File: internal/npm_install/npm_amd_bundle.bzl [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+"""Node package amd bundling
+
+For use by yarn_install and npm_install. Not meant to be part of the public API.
+"""
+
+load("@build_bazel_rules_nodejs//internal/common:node_module_info.bzl", "NodeModuleSources", "collect_node_modules_aspect")
+load("//internal/common:expand_into_runfiles.bzl", "expand_path_into_runfiles")
+
+def _filter_js_inputs(all_inputs):
+    # Note: make sure that "all_inputs" is not a depset.
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/npm_install/npm_amd_bundle.bzl]: I simplified and removed this function; can just filter inline instead in this case as its only `.js` files needed

**Comment by jbedard** [comment]
Does this also (partially?) solve https://github.com/bazelbuild/rules_nodejs/issues/436 ?

**Comment by gregmagolan** [comment]
In theory yes. It would provide a browseiry generated named UMD bundle for any npm package. Whether it will work for all npm packages remains to be seen.

**Comment by gregmagolan** [comment]
Also partially fixes https://github.com/bazelbuild/rules_nodejs/pull/725 tho it doesn't make the support for commonjs in ts_devserver and ts_web_test_suite automatic yet. One step at a time.

File: internal/npm_install/browserify-wrapped.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexeagle) [reply internal/npm_install/browserify-wrapped.js]: spawning a process is fast, I guess it's a micro-optimization to try to stay in one VM.
It does mean that options like max heap size are applied to this wrapper program and not to the library which makes me worry a little about child processes

File: third_party/github.com/browserify/browserify/BUILD.bazel (hunk 3 above)
Comment (gregmagolan) [reply third_party/github.com/browserify/browserify/BUILD.bazel]: 👍 :

File: third_party/github.com/jhermsmeier/browserify-named-amd/BUILD.bazel (hunk 4 above)
Comment (gregmagolan) [reply third_party/github.com/jhermsmeier/browserify-named-amd/BUILD.bazel]: 👍 :

**Comment by gregmagolan** [comment]
Note: using the "main" entry point as is done now may not be correct and may not work for all npm packages. http://2ality.com/2017/04/setting-up-multi-platform-packages.html.
It may be better to add the esmify browserify plugin and prefer the "browser" and then "module" and then "main" entry points. Some more thought and testing required and will leave that for another PR.

**Pull request merged**